r2/r2/lib/base.py

```diff
@@ -103,8 +112,8 @@ def __call__(self, environ, start_response):
         if cdn_ip:
             request.ip = cdn_ip
             request.via_cdn = True
-        elif g.trust_local_proxies and forwarded_for and is_local_address(remote_addr):
-            request.ip = forwarded_for.split(',')[-1]
+        elif g.trust_local_proxies and forwarded_for and is_trusted_proxy(remote_addr):
+            request.ip = forwarded_for.split(',')[0].strip()
```

Comment:
I think there's some confusion here or we may be talking at cross purposes.

> The reason we pull the last IP is that it's the only one we can trust

As you say, XFF header can't be trusted unless the immediate peer (`environ['REMOTE_ADDR]`) is trusted (and, implicitly, this instance trusts the chain of IPs that the peer trusts etc).  

The new INI kw `trusted_proxy_ips` provides the list of trusted immediate peers.  I only check the immediate peer on the assumption that the trusted peer itself either replaces the XFF header entirely or trusts an upstream XFF header if present, but I suppose I could adapt `is_trusted_proxy()` to check all the IPs in the XFF chain if desired.  It seems superfluous, though, and would incur additional load.

Anyway, once the trustworthiness of the peer has been established, the XFF header can be considered valid.  At this point, the objective is to assign to `request.ip` the IP of the original requestor (just as original line 109 uses the peer's IP when the peer is not trusted or there's no XFF header).

If XFF contains only one IP (as would be the case in default dev installs), both the first and last IP is that of the original requestor.  If, however, there is more than one IP in the XFF header (such as happens when Varnish cache is used in lieu of HAProxy), the first IP is the original requestor and every subsequent IP will be the IPs of machines in the trust chain with the last IP being the most recent proxy (and so should match `environ['REMOTE_ADDR]` if it is appending to the XFF as it should).

That's why the correct thing to do is to take the first IP rather than any of the subsequent IPs.